packages/marketprice/src/alarms/mod.rs

```diff
@@ -77,136 +75,241 @@ impl<'a> Prefixer<'a> for AlarmStore {
-            .idx
-            .0
-            .sub_prefix(C::TICKER.into())
-            .range(
-                storage,
-                Some(Bound::exclusive((price.0.amount(), Addr::unchecked("")))),
-                None,
-                Order::Ascending,
+        Self::add_alarm_internal(
+            &self.alarms_below,
+            &mut *self.storage,
+            subscriber,
+            &AlarmStore::new(&alarm),
+        )
+    }
+
+    pub fn add_alarm_above_or_equal<C, BaseC>(
+        &mut self,
+        subscriber: Addr,
+        alarm: Price<C, BaseC>,
+    ) -> Result<(), AlarmError>
+    where
+        C: Currency,
+        BaseC: Currency,
+    {
+        Self::add_alarm_internal(
+            &self.alarms_above_or_equal,
+            &mut *self.storage,
+            subscriber,
+            &AlarmStore::new(&alarm),
+        )
+    }
+
+    pub fn remove(&mut self, addr: Addr) -> Result<(), AlarmError> {
+        self.alarms_below
+            .remove(&mut *self.storage, addr.clone())
+            .and_then(|()| self.alarms_above_or_equal.remove(&mut *self.storage, addr))
+            .map_err(Into::into)
+    }
+
+    pub fn out_for_delivery(&mut self, subscriber: Addr) -> Result<(), AlarmError> {
+        let below: AlarmStore = self.alarms_below.load(&*self.storage, subscriber.clone())?;
+
+        self.alarms_below
+            .replace(&mut *self.storage, subscriber.clone(), None, Some(&below))?;
+
+        let above: Option<AlarmStore> = self
+            .alarms_above_or_equal
+            .may_load(&*self.storage, subscriber.clone())?;
+
+        if let Some(above) = &above {
+            self.alarms_below
+                .replace(&mut *self.storage, subscriber.clone(), None, Some(above))?;
+        }
+
+        self.in_delivery
+            .push_back(
+                &mut *self.storage,
+                &AlarmWithSubscriber {
+                    subscriber,
+                    below,
+                    above,
+                },
             )
+            .map_err(Into::into)
+    }
+
+    pub fn last_delivered(&mut self) -> Result<(), AlarmError> {
+        self.in_delivery
+            .pop_front(&mut *self.storage)
+            .map_err(Into::into)
+            .and_then(|maybe_alarm: Option<AlarmWithSubscriber>| {
+                if maybe_alarm.is_some() {
+                    Ok(())
+                } else {
+                    Err(AlarmError::EmptyAlarmsInDeliveryQueue(String::from(
+                        "Received success reply status",
+                    )))
+                }
+            })
+    }
+
+    pub fn last_failed(&mut self) -> Result<(), AlarmError> {
+        self.in_delivery
+            .pop_front(&mut *self.storage)
+            .map_err(Into::into)
+            .and_then(|maybe_alarm: Option<AlarmWithSubscriber>| {
+                maybe_alarm.ok_or(AlarmError::EmptyAlarmsInDeliveryQueue(String::from(
+                    "Received failure reply status",
+                )))
```

Comment:
wonder if there would be any gains in using the lazy `ok_or_else` ... if in doubt check that out on https://godbolt.org/

Reply:
Usually there is. Clippy used to complain about it, but I suppose it was when in pedantic mode. Will replace it it with `ok_or_else` because it will skip one unnecessary allocation as far as I am aware.